src/mscorlib/src/System/Runtime/InteropServices/NativeLibrary.cs

```diff
@@ -0,0 +1,365 @@
+
+            if (charSet == CharSet.Ansi)
+            {
+                Debug.Assert(retVal == null);
+                retVal = CreateDelegateForSymbolByName<TDelegate>(name + "A");
+            }
+            else
+            {
+                // On Unicode only (not ANSI), a method with a 'W' suffix takes precedence
+                // over a method with an exact name match when name mangling is enabled.
+                retVal = CreateDelegateForSymbolByName<TDelegate>(name + "W") ?? retVal;
+            }
+
+#if (BIT32 && !ARM)
+            // On x86 only, look for __stdcall mangled names as a last resort.
+            // They'll be of the form "_Name@x", where x is the stack size in bytes of the method arguments.
+
+            if (retVal == null && callingConvention == CallingConvention.StdCall)
+            {
+                Debug.Assert(!exactSpelling);
+                string stdcallMangledName = FormattableString.Invariant($"_{name}@{(uint)Marshal.NumParamBytes(delegateInvokeMethod, isForStdCallDelegate: true):D}");
+                retVal = CreateDelegateForSymbolByName<TDelegate>(stdcallMangledName);
+            }
+#endif
+
+            // This could still be null if we had no match, but there's nothing left for us to try.
+            return retVal;
+        }
+
+        // Looks up a symbol by name (caller is responsible for mangling), and if it's found returns a delegate for it.
+        // Returns null if the symbol is not found.
+        private TDelegate CreateDelegateForSymbolByName<TDelegate>(string symbolName) where TDelegate : class
+        {
+            Debug.Assert(((RuntimeType)typeof(TDelegate)).IsDelegate());
+
+            IntPtr symbolAddress = GetProcAddress(_hInstance, symbolName);
+            return (symbolAddress != IntPtr.Zero)
+                ? Marshal.GetDelegateForFunctionPointer<TDelegate>(symbolAddress)
+                : null;
+        }
+
+        // Looks up a symbol by ordinal and if it's found returns a delegate for it.
+        // Returns null if the symbol is not found.
+        private TDelegate CreateDelegateForSymbolByOrdinal<TDelegate>(ushort ordinal) where TDelegate : class
+        {
+            Debug.Assert(((RuntimeType)typeof(TDelegate)).IsDelegate());
+
+            IntPtr symbolAddress = GetProcAddress(_hInstance, (IntPtr)ordinal);
+            return (symbolAddress != IntPtr.Zero)
+                ? Marshal.GetDelegateForFunctionPointer<TDelegate>(symbolAddress)
+                : null;
+        }
+
+        private static void ParseCharsetAndCallingConvention(Type delegateType, out CharSet charSet, out CallingConvention callingConvention)
+        {
+            Debug.Assert(delegateType is RuntimeType);
+            Debug.Assert(((RuntimeType)delegateType).IsDelegate());
+
+            charSet = default;
+            callingConvention = default;
+
+            var attr = delegateType.GetCustomAttribute<UnmanagedFunctionPointerAttribute>(inherit: false);
```

Comment:
I do not think we should be duplicating the UnmanagedFunctionPointerAttribute parsing logic here. We should just use the one we have in unmanaged runtime.

Reply:
(Same for other pieces of logic that are duplicated in the unmanaged runtime.)

Reply:
I can remove `ParseCharsetAndCallingConvention` by reusing existing `PInvokeStaticSigInfo` logic in dllimport.h. However, I don't think I can reuse existing unmanaged `FindEntryPoint` logic without introducing a significant refactoring, and others I spoke with offline offered that the current proposal of rewriting the logic in managed code would be much less risky.

Reply:
I does not have to reusing the code if it is unnatural. But I think having a method with same structure as `NDirectMethodDesc::FindEntryPoint` right next to `NDirectMethodDesc::FindEntryPoint` to handle this case would be better.